engineapi/proxy.go

```diff
@@ -1,168 +1,421 @@
+
+	go wait.Until(h.resync, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (h *EngineClientProxyHandler) resync() {
+	for {
+		handledPools := h.getHandledEngineClientProxyPools()
+		for instanceManagerImage, pool := range handledPools {
+			im, err := h.ds.GetInstanceManagerRO(pool.instanceManagerName)
+			if err != nil {
+				if !datastore.ErrorIsNotFound(err) {
+					h.logger.WithError(err).Warnf("Failed to get %v instance manager for engine client proxy", pool.instanceManagerName)
+					continue
+				}
+
+				h.logger.Warnf("%v instance manager no longer exist for engine client proxy", pool.instanceManagerName)
+				for _, client := range pool.clients {
+					engineClientProxy := client.getEngineClientProxy()
+					log := engineClientProxy.getLogger()
+					log.Debug("Cleanup engine client proxy")
+
+					engineClientProxy.setNextGC()
+				}
+				continue
+			}
+
+			for clientUUID, client := range pool.clients {
+				engineClientProxy := client.getEngineClientProxy()
+				if !engineClientProxy.shouldResync() {
+					continue
+				}
+
+				log := engineClientProxy.getLogger()
+
+				if engineClientProxy.shouldDelete() {
+					log.Debug("Deleting engine client proxy")
+					engineClientProxy.setNextGC()
+
+					if err := h.NewEngineClientProxy(im); err != nil {
+						log.WithError(err).Debug("Failed to replace with new engine client proxy")
+					}
+
+					engineClientProxy.setNextResync(h.resyncPeriod)
+					break
+				}
+
+				h.gc(instanceManagerImage, clientUUID)
+
+				if engineClientProxy.shouldCreate(im) {
+					log.Debug("Creating engine client proxy")
+
+					if err := engineClientProxy.start(im, h.logger, h.ds, h.stopCh); err != nil {
+						log.Warn("Failed to start engine client proxy")
+					}
+				}
+
+				engineClientProxy.setNextResync(h.resyncPeriod)
+			}
+
+			exceedMaxConnection := true
+			for _, client := range pool.clients {
+				engineClientProxy := client.getEngineClientProxy()
+				if engineClientProxy.getInCall() >= MaxProxyClientConcurrentInUse {
+					continue
+				}
+
+				exceedMaxConnection = false
+				break
+			}
+			if len(pool.clients) != 0 && exceedMaxConnection {
+				log := getLoggerForProxyHandler(h.logger, im)
+				log.Debugf("Exceed engine client proxy max concurrent use limit:%v", MaxProxyClientConcurrentInUse)
+
+				if err := h.NewEngineClientProxy(im); err != nil {
+					log.WithError(err).Debug("Failed to add more engine client proxy")
+				}
+			}
+		}
+	}
+}
+
+func (h *EngineClientProxyHandler) gc(instanceManagerImage string, clientUUID string) {
+	client, exist := h.getEngineClientProxy(instanceManagerImage, clientUUID)
+	if !exist {
+		h.logger.Debugf("Engine client proxy(%v) for %v is already gone", clientUUID, instanceManagerImage)
+		return
+	}
+
+	log := client.getLogger()
+
+	nextGC := client.getNextGC()
+	now := time.Now()
+	if nextGC == nil || now.Before(*nextGC) {
+		return
+	}
+
+	msg := ""
+	if client.getInCall() != 0 {
```

Comment:
note: We need to ensure if the client has been marked GC, and after that, it should be not used anymore.

Reply:
Logic in `checkClient`.